File: examples/showcase/travel/capabilities.py
"""Travel booking capabilities — ANIP capability declarations and handlers."""
import time
import uuid

from anip_service import Capability, InvocationContext, ANIPError
from anip_core import (
    BindingRequirement, CapabilityDeclaration, CapabilityInput, CapabilityOutput,
    CapabilityRequirement, Cost, CostCertainty, CrossServiceHints, FinancialCost,
    ObservabilityContract, ResponseMode, ServiceCapabilityRef, SessionInfo, SideEffect,
    SideEffectType,
)
import data

# ---------------------------------------------------------------------------
# 1. search_flights — read-only, streaming response mode
# ---------------------------------------------------------------------------

_SEARCH_DECL = CapabilityDeclaration(
    name="search_flights",
    description="Search available flights by origin and destination",
    contract_version="1.0",
    inputs=[
        CapabilityInput(name="origin", type="airport_code", description="Departure airport (IATA code)"),
        CapabilityInput(name="destination", type="airport_code", description="Arrival airport (IATA code)"),
    ],
    output=CapabilityOutput(
        type="flight_list",
        fields=["flight_number", "origin", "destination", "departure_time", "arrival_time", "price", "stops", "quote_id"],
    ),
    side_effect=SideEffect(type=SideEffectType.READ, rollback_window="not_applicable"),
    minimum_scope=["travel.search"],
    cost=Cost(certainty=CostCertainty.FIXED, financial=None, compute={"latency_p50": "150ms", "tokens": 400}),
    session=SessionInfo(),
    response_modes=[ResponseMode.UNARY, ResponseMode.STREAMING],
    observability=ObservabilityContract(
        logged=True, retention="P90D",
        fields_logged=["capability", "parameters", "result_count"],
        audit_accessible_by=["delegation.root_principal"],
    ),
    # Cross-service hints (illustrative — this is a single-service showcase)
    cross_service=CrossServiceHints(
        handoff_to=[ServiceCapabilityRef(service="travel-booking", capability="book_flight")],
    ),
)


def _handle_search(ctx: InvocationContext, params: dict) -> dict:
    origin = params.get("origin")
    destination = params.get("destination")
    if not origin or not destination:
        raise ANIPError("invalid_parameters", "origin and destination are required")

    flights = data.search_flights(origin, destination)
    return {
        "flights": [
            {
                "flight_number": f.flight_number,
                "origin": f.origin,
                "destination": f.destination,
                "departure_time": f.departure_time,
                "arrival_time": f.arrival_time,
                "price": f.price,
                "currency": f.currency,
                "stops": f.stops,
                "quote_id": {
                    "id": f"qt-{uuid.uuid4().hex[:8]}-{int(time.time())}",
                    "price": f.price,
                    "issued_at": int(time.time()),
                },
            }
            for f in flights
        ],
        "count": len(flights),
    }


search_flights = Capability(declaration=_SEARCH_DECL, handler=_handle_search)


# ---------------------------------------------------------------------------
# 2. check_availability — read-only
# ---------------------------------------------------------------------------

_AVAIL_DECL = CapabilityDeclaration(
    name="check_availability",
    description="Check seat availability and price for a specific flight",
    contract_version="1.0",
    inputs=[
        CapabilityInput(name="flight_number", type="string", description="Flight number to check"),
    ],
    output=CapabilityOutput(
        type="availability_info",
        fields=["available", "seats_remaining", "flight_number", "price"],
    ),
    side_effect=SideEffect(type=SideEffectType.READ, rollback_window="not_applicable"),
    minimum_scope=["travel.search"],
    cost=Cost(certainty=CostCertainty.FIXED, financial=None, compute={"latency_p50": "50ms", "tokens": 200}),
    session=SessionInfo(),
    observability=ObservabilityContract(
        logged=True, retention="P90D",
        fields_logged=["capability", "parameters", "result"],
        audit_accessible_by=["delegation.root_principal"],
    ),
)


def _handle_availability(ctx: InvocationContext, params: dict) -> dict:
    flight_number = params.get("flight_number")
    if not flight_number:
        raise ANIPError("invalid_parameters", "flight_number is required")

    result = data.check_availability(flight_number)
    if not result["available"] and result.get("reason") == "flight_not_found":
        raise ANIPError("capability_unavailable", f"Flight {flight_number} not found")
    return result


check_availability = Capability(declaration=_AVAIL_DECL, handler=_handle_availability)


# ---------------------------------------------------------------------------
# 3. book_flight — irreversible, financial cost, requires search_flights
# ---------------------------------------------------------------------------

_BOOK_DECL = CapabilityDeclaration(
    name="book_flight",
    description="Book a confirmed flight reservation",
    contract_version="1.0",
    inputs=[
        CapabilityInput(name="flight_number", type="string", description="Flight to book"),
        CapabilityInput(name="quote_id", type="object", required=True, description="Priced quote from search_flights containing id, price, and issued_at"),
        CapabilityInput(
            name="passengers", type="integer", required=False, default=1,
            description="Number of passengers",
        ),
    ],
    output=CapabilityOutput(
        type="booking_confirmation",
        fields=["booking_id", "flight_number", "departure_time", "total_cost"],
    ),
    side_effect=SideEffect(type=SideEffectType.IRREVERSIBLE, rollback_window="none"),
    minimum_scope=["travel.book"],
    cost=Cost(
        certainty=CostCertainty.ESTIMATED,
        financial=FinancialCost(
            currency="USD",
            range_min=280,
            range_max=500,
            typical=420,
        ),
        determined_by="search_flights",
        compute={"latency_p50": "1s", "tokens": 1000},
    ),
    requires=[
        CapabilityRequirement(
            capability="search_flights",
            reason="must select from available flights before booking",
        ),
    ],
    requires_binding=[
        BindingRequirement(
            type="quote",
            field="quote_id",
            source_capability="search_flights",
            max_age="PT15M",
        ),
    ],
    refresh_via=["search_flights"],
    session=SessionInfo(),
    observability=ObservabilityContract(
        logged=True, retention="P90D",
        fields_logged=["capability", "delegation_chain", "parameters", "result", "cost_actual"],
        audit_accessible_by=["delegation.root_principal"],
    ),
    # Cross-service hints (illustrative — this is a single-service showcase)
    cross_service=CrossServiceHints(
        refresh_via=[ServiceCapabilityRef(service="travel-search", capability="search_flights")],
    ),
)


def _handle_book(ctx: InvocationContext, params: dict) -> dict:
    flight_number = params.get("flight_number")
    passengers = params.get("passengers", 1)
    quote_id = params.get("quote_id")
    if isinstance(quote_id, dict):
        booking_ref = quote_id.get("id", "")
        quoted_price = quote_id.get("price")
    else:
        booking_ref = quote_id  # backward compat for plain strings
        quoted_price = None

    if not flight_number:
        raise ANIPError("invalid_parameters", "flight_number is required")

    flight = data.get_flight(flight_number)
    if flight is None:
        raise ANIPError(
            "capability_unavailable",
            f"Flight {flight_number} not found",
            resolution={
                "action": "search_flights",
                "recovery_class": "refresh_then_retry",
                "recovery_target": {
                    "kind": "refresh",
                    "target": {"service": "travel-booking", "capability": "search_flights"},
                    "continuity": "same_task",
                    "retry_after_target": True,
                },
            },
        )

    if flight.seats_available < passengers:
        raise ANIPError(
            "capability_unavailable",
            f"Only {flight.seats_available} seats left on {flight_number}",
            resolution={
                "action": "check_availability",
                "recovery_class": "refresh_then_retry",
                "recovery_target": {
                    "kind": "refresh",
                    "target": {"service": "travel-booking", "capability": "check_availability"},
                    "continuity": "same_task",
                    "retry_after_target": True,
                },
            },
        )

    booking = data.create_booking(
        flight=flight,
        passengers=passengers,
        booked_by=ctx.subject,
        on_behalf_of=ctx.root_principal,
    )

    ctx.set_cost_actual({"financial": {"amount": booking.total_cost, "currency": booking.flight.currency}})

    return {
        "booking_id": booking.booking_id,
        "flight_number": booking.flight.flight_number,
        "departure_time": booking.flight.departure_time,
        "total_cost": booking.total_cost,
        "currency": booking.flight.currency,
    }


book_flight = Capability(declaration=_BOOK_DECL, handler=_handle_book)


# ---------------------------------------------------------------------------
# 4. cancel_booking — transactional, rollback window PT24H
# ---------------------------------------------------------------------------

_CANCEL_DECL = CapabilityDeclaration(
    name="cancel_booking",
    description="Cancel an existing flight booking",
    contract_version="1.0",
    inputs=[
        CapabilityInput(name="booking_id", type="string", description="Booking ID to cancel"),
    ],
    output=CapabilityOutput(
        type="cancellation_confirmation",
        fields=["booking_id", "status", "refund_amount"],
    ),
    side_effect=SideEffect(type=SideEffectType.TRANSACTIONAL, rollback_window="PT24H"),
    minimum_scope=["travel.book"],
    cost=Cost(certainty=CostCertainty.FIXED, financial=None, compute={"latency_p50": "500ms", "tokens": 300}),
    session=SessionInfo(),
    observability=ObservabilityContract(
        logged=True, retention="P90D",
        fields_logged=["capability", "delegation_chain", "parameters", "result"],
        audit_accessible_by=["delegation.root_principal"],
    ),
)


def _handle_cancel(ctx: InvocationContext, params: dict) -> dict:
    booking_id = params.get("booking_id")
    if not booking_id:
        raise ANIPError("invalid_parameters", "booking_id is required")

    try:
        booking = data.cancel_booking(booking_id)
    except KeyError:
        raise ANIPError(
            "capability_unavailable",
            f"Booking {booking_id} not found",
            resolution={
                "action": "revalidate_booking_id",
                "recovery_class": "revalidation",
                "recovery_target": {
                    "kind": "revalidation",
                    "continuity": "same_task",
                    "retry_after_target": True,
                },
            },
        )
    except ValueError as exc:
        raise ANIPError("invalid_parameters", str(exc))

    return {
        "booking_id": booking.booking_id,
        "status": booking.status,
        "refund_amount": booking.total_cost,
        "currency": booking.flight.currency,
    }


cancel_booking = Capability(declaration=_CANCEL_DECL, handler=_handle_cancel)

File: packages/go/examples/flights/capabilities.go
package main

import (
	"crypto/rand"
	"fmt"
	"time"

	"github.com/anip-protocol/anip/packages/go/core"
	"github.com/anip-protocol/anip/packages/go/service"
)

func floatPtr(f float64) *float64 { return &f }

func randomHex(n int) string {
	b := make([]byte, n)
	rand.Read(b)
	return fmt.Sprintf("%x", b)
}

// SearchFlights returns the search_flights capability definition.
func SearchFlights() service.CapabilityDef {
	return service.CapabilityDef{
		Declaration: core.CapabilityDeclaration{
			Name:            "search_flights",
			Description:     "Search available flights by origin, destination, and date",
			ContractVersion: "1.0",
			Inputs: []core.CapabilityInput{
				{Name: "origin", Type: "airport_code", Required: true, Description: "Departure airport"},
				{Name: "destination", Type: "airport_code", Required: true, Description: "Arrival airport"},
				{Name: "date", Type: "date", Required: true, Description: "Travel date (YYYY-MM-DD)"},
			},
			Output: core.CapabilityOutput{
				Type:   "flight_list",
				Fields: []string{"flight_number", "departure_time", "arrival_time", "price", "stops", "quote_id"},
			},
			SideEffect: core.SideEffect{
				Type:           "read",
				RollbackWindow: "not_applicable",
			},
			MinimumScope:  []string{"travel.search"},
			ResponseModes: []string{"unary"},
			RefreshVia:    []string{},
			VerifyVia:     []string{},
		},
		Handler: handleSearchFlights,
	}
}

// BookFlight returns the book_flight capability definition.
func BookFlight() service.CapabilityDef {
	return service.CapabilityDef{
		Declaration: core.CapabilityDeclaration{
			Name:            "book_flight",
			Description:     "Book a confirmed flight reservation",
			ContractVersion: "1.0",
			Inputs: []core.CapabilityInput{
				{Name: "flight_number", Type: "string", Required: true, Description: "Flight to book"},
				{Name: "date", Type: "date", Required: true, Description: "Travel date (YYYY-MM-DD)"},
				{Name: "passengers", Type: "integer", Required: false, Default: 1, Description: "Number of passengers"},
				{Name: "quote_id", Type: "object", Required: false, Description: "Priced quote from search_flights"},
			},
			Output: core.CapabilityOutput{
				Type:   "booking_confirmation",
				Fields: []string{"booking_id", "flight_number", "departure_time", "total_cost"},
			},
			SideEffect: core.SideEffect{
				Type:           "irreversible",
				RollbackWindow: "none",
			},
			MinimumScope: []string{"travel.book"},
			Cost: &core.Cost{
				Certainty: "estimated",
				Financial: &core.FinancialCost{
					Currency: "USD",
					RangeMin: floatPtr(280),
					RangeMax: floatPtr(500),
				},
				DeterminedBy: "search_flights",
			},
			Requires: []core.CapabilityRequirement{
				{
					Capability: "search_flights",
					Reason:     "must select from available flights before booking",
				},
			},
			RequiresBinding: []core.BindingRequirement{
				{
					Type:             "quote",
					Field:            "quote_id",
					SourceCapability: "search_flights",
					MaxAge:           "PT15M",
				},
			},
			ControlRequirements: []core.ControlRequirement{},
			RefreshVia:          []string{"search_flights"},
			VerifyVia:           []string{},
			ResponseModes:       []string{"unary"},
		},
		Handler: handleBookFlight,
	}
}

// --- Flight data ---

type flight struct {
	FlightNumber  string  `json:"flight_number"`
	Origin        string  `json:"origin"`
	Destination   string  `json:"destination"`
	Date          string  `json:"date"`
	DepartureTime string  `json:"departure_time"`
	ArrivalTime   string  `json:"arrival_time"`
	Price         float64 `json:"price"`
	Currency      string  `json:"currency"`
	Stops         int     `json:"stops"`
}

var flightInventory = []flight{
	{"AA100", "SEA", "SFO", "2026-03-10", "08:00", "10:15", 420.00, "USD", 0},
	{"UA205", "SEA", "SFO", "2026-03-10", "11:30", "13:45", 380.00, "USD", 0},
	{"DL310", "SEA", "SFO", "2026-03-10", "14:00", "18:30", 280.00, "USD", 1},
	{"AA101", "SEA", "SFO", "2026-03-11", "08:00", "10:15", 310.00, "USD", 0},
	{"UA450", "SEA", "LAX", "2026-03-10", "09:00", "11:30", 350.00, "USD", 0},
	{"DL520", "SFO", "JFK", "2026-03-12", "06:00", "14:30", 580.00, "USD", 0},
}

func searchFlightData(origin, destination, date string) []flight {
	var results []flight
	for _, f := range flightInventory {
		if f.Origin == origin && f.Destination == destination && f.Date == date {
			results = append(results, f)
		}
	}
	return results
}

func findFlight(flightNumber, date string) *flight {
	for _, f := range flightInventory {
		if f.FlightNumber == flightNumber && f.Date == date {
			return &f
		}
	}
	return nil
}

// --- Handlers ---

func handleSearchFlights(ctx *service.InvocationContext, params map[string]any) (map[string]any, error) {
	origin, _ := params["origin"].(string)
	destination, _ := params["destination"].(string)
	date, _ := params["date"].(string)

	if origin == "" || destination == "" || date == "" {
		return nil, core.NewANIPError("invalid_parameters", "origin, destination, and date are all required")
	}

	flights := searchFlightData(origin, destination, date)
	flightMaps := make([]map[string]any, len(flights))
	for i, f := range flights {
		flightMaps[i] = map[string]any{
			"flight_number":  f.FlightNumber,
			"departure_time": f.DepartureTime,
			"arrival_time":   f.ArrivalTime,
			"price":          f.Price,
			"currency":       f.Currency,
			"stops":          f.Stops,
			"quote_id": map[string]any{
				"id":        fmt.Sprintf("qt-%s-%d", randomHex(4), time.Now().Unix()),
				"price":     f.Price,
				"issued_at": time.Now().Unix(),
			},
		}
	}

	return map[string]any{
		"flights": flightMaps,
		"count":   len(flights),
	}, nil
}

func handleBookFlight(ctx *service.InvocationContext, params map[string]any) (map[string]any, error) {
	flightNumber, _ := params["flight_number"].(string)
	date, _ := params["date"].(string)
	passengers := 1
	if p, ok := params["passengers"].(float64); ok {
		passengers = int(p)
	}

	// Accept quote_id; extract price override if provided.
	var quotedPrice float64
	if qid, ok := params["quote_id"].(map[string]any); ok {
		if p, ok := qid["price"].(float64); ok {
			quotedPrice = p
		}
	}

	if flightNumber == "" || date == "" {
		return nil, core.NewANIPError("invalid_parameters", "flight_number and date are required")
	}

	f := findFlight(flightNumber, date)
	if f == nil {
		return nil, core.NewANIPError(core.FailureUnavailable, fmt.Sprintf("flight %s on %s not found", flightNumber, date))
	}

	price := f.Price
	if quotedPrice > 0 {
		price = quotedPrice
	}
	totalCost := price * float64(passengers)

	// Generate booking ID.
	b := make([]byte, 4)
	rand.Read(b)
	bookingID := fmt.Sprintf("BK-%X", b)

	// Track actual cost.
	ctx.SetCostActual(&core.CostActual{
		Financial: &core.FinancialCost{
			Currency: f.Currency,
			Amount:   &totalCost,
		},
	})

	return map[string]any{
		"booking_id":     bookingID,
		"flight_number":  f.FlightNumber,
		"departure_time": f.DepartureTime,
		"total_cost":     totalCost,
		"currency":       f.Currency,
	}, nil
}
